include/openthread/border_agent.h

```diff
@@ -123,6 +123,36 @@ bool otBorderAgentIsActive(otInstance *aInstance);
  */
 uint16_t otBorderAgentGetUdpPort(otInstance *aInstance);
 
+/**
+ * This callback informs the application the changes of the state of the MeshCoP service.
+ *
+ * In specific, the 'state' includes the MeshCoP TXT data originated from the Thread network and whether the
+ * Border Agent is Active (which can be obtained by `otBorderAgentIsActive`).
+ *
+ * @param[in] aTxtData  A pointer to the encoded MeshCoP TXT data originated from the Thread network.
+ * @param[in] aLength   The length of the encoded TXT data.
+ * @param[in] aContext  A pointer to application-specific context.
+ */
+typedef void (*otBorderAgentStateChangedCallback)(const uint8_t *aTxtData, uint16_t aLength, void *aContext);
+
+/**
+ * Sets the callback function used by the Border Agent to notify any changes on the state of the MeshCoP service.
+ *
+ * The callback will be invoked when the 'Is Active' state of the Border Agent or the MeshCoP service TXT data values
+ * changes. For example, it will be invoked when the network name or the extended PAN ID changes and pass the updated
+ * encoded TXT data to the application layer.
+ *
+ * This callback will be invoked once right after this API is called to provide initial states of the MeshCoP service
+ * to the application.
+ *
+ * @param[in] aInstance  A pointer to an OpenThread instance.
+ * @param[in] aCallback  The callback to be invoked when there are any changes of the MeshCoP service.
+ * @param[in] aContext   A pointer to application-specific context.
+ */
+void otBorderAgentSetStateChangedCallback(otInstance                       *aInstance,
+                                          otBorderAgentStateChangedCallback aCallback,
+                                          void                             *aContext);
```

Comment:
How about  renaming the callback to indicate its intended use more clearly? Since it not only related to BA state changes but any of paramter used by MeshCoP Service (in TXT record)?

Maybe something like `otBorderAgentSetMeshCoPServiceChangedCallback` (and `otBorderAgentMeshCoPServiceChangedCallback`

Reply:
Sounds good! Updated.